state/accumulator.go

```diff
@@ -412,11 +415,40 @@ func (a *Accumulator) Accumulate(txn *sqlx.Tx, userID, roomID string, prevBatch
 			if ev.NID > latestNID {
 				latestNID = ev.NID
 			}
+			// if this is a redaction, try to redact the referenced event (best effort)
+			if !ev.IsState && ev.Type == "m.room.redaction" {
+				// look for top-level redacts then content.redacts (room version 11+)
+				redactsEventID := parsedEv.Get("redacts").Str
+				if redactsEventID == "" {
+					redactsEventID = parsedEv.Get("content.redacts").Str
+				}
+				if redactsEventID != "" {
+					redactTheseEventIDs = append(redactTheseEventIDs, redactsEventID)
+				}
+			}
 			newEvents = append(newEvents, ev)
 			timelineNIDs = append(timelineNIDs, ev.NID)
 		}
 	}
 
+	// if we are going to redact things, we need the room version to know the redaction algorithm
+	// so pull it out once now.
+	var roomVersion string
+	if len(redactTheseEventIDs) > 0 {
+		createEventJSON, err := a.eventsTable.SelectCreateEvent(txn, roomID)
+		if err != nil {
+			return 0, nil, fmt.Errorf("SelectCreateEvent: %s", err)
+		}
+		roomVersion = gjson.GetBytes(createEventJSON, "content.room_version").Str
+		if roomVersion == "" {
+			// Defaults to "1" if the key does not exist.
+			roomVersion = "1"
+		}
+		if err = a.eventsTable.Redact(txn, roomVersion, redactTheseEventIDs); err != nil {
+			return 0, nil, err
+		}
```

Comment:
If something goes wrong here this will stop the accumulation in its tracks. Maybe that's for the best though?

Reply:
It means we will retry as Accumulate is a retryable function. This is good if `err` is because of a temporary DB connection blip. This is bad if it's a data error as it means we'll never make forward progress.